nixos/sudo: Add support for sudo-rs

nicoonoclaste opened the pull request:
## Description of changes

- nixos/sudo:
  - [x] Many, many refactors
  - [x] Check `sudoers`' syntax using the configured package
  - [x] Added `defaultOptions` config option, that can override behaviour incompatible with sudo-rs
- [x] nixos/terminfo: Added option not to add extra sudo config which is incompatible with sudo-rs
- [x] Quickly looked through other NixOS modules for potential incompatibilities with sudo-rs,
      added an assertion in the only one I found (`x2goserver`)
- [x] Add NixOS test for `sudo-rs`

See tracking issue #253465

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Module updates) Added a release notes entry if the change is significant
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nicoonoclaste committed:
nixos/sudo: Split up `configFile` into individual sections

nicoonoclaste committed:
nixos/sudo: Don't include empty sections

This makes the generated sudoers a touch easier to read.

nicoonoclaste committed:
nixos/sudo: Only keep SSH_AUTH_SOCK if used for authentication

This will make compatibility with `sudo-rs` easier.

nicoonoclaste committed:
nixos/terminfo: Add config option not to add extra sudo config

This will be necessary for compatibility with `sudo-rs`.

nicoonoclaste committed:
nixos/sudo: Refactor option definitions

nicoonoclaste committed:
nixos/sudo: Refactor checks for Todd C. Miller's implemetation

nicoonoclaste committed:
nixos/sudo: Drop useless `lib.` qualifiers

Also normalise indentation for `mdDoc` to what's prevalent in this file.

nicoonoclaste committed:
nixos/sudo: Handle `root`'s default rule through `extraRules`

This makes things more uniform, and simplifies compatibility with sudo-rs.

Moreover, users can not inject rules before this if they need to.

nicoonoclaste committed:
nixos/sudo: Make the default rules' options configurable

nicoonoclaste committed:
nixos/sudo: Drop the sudoers comment for `extraRules`

All rules are now handled through `extraRules`,
and it is never empty so `optionalString` isn't needed either.

nicoonoclaste committed:
nixos/sudo: Check syntax using the configured package

This is preferable even for regular `sudo`, but will ensure the check is useful
when using `sudo-rs` in the future.

Also, dropped antediluvian comment about the syntax check being disabled,
when it was clearly not commented out:
  - introduced in 2007, commit 6d65f0ae03ae14f3e978d89959253d9a8f5e0ec1;
  - reverted in 2014, commit e68a5b265a96134243a1572f43dfc4ff75dd082b,
    but without ammending the comments.

nicoonoclaste committed:
nixos/{sudo, terminfo}: Adjust defaults for compatibility with `sudo-rs`

emilylange commented:
> Apparently the CI just fails on draft PRs:

No, that's due to a merge conflict in `nixos/doc/manual/release-notes/rl-2311.section.md` you will have to fix (by rebasing).

nicoonoclaste committed:
nixos/sudo: Add tests for sudo-rs too

Duplicated sudo's testsuite for now, as its maintainer does not with
to collaborate on testing effors; see #253876.

Environment-related tests were removed, as sudo-rs does not support
`(NO)SETENV` yet; see memorysafety/sudo-rs#760

mkg20001 commented on nixos/tests/all-tests.nix:
```diff
@@ -740,6 +740,7 @@ in {
   strongswan-swanctl = handleTest ./strongswan-swanctl.nix {};
   stunnel = handleTest ./stunnel.nix {};
   sudo = handleTest ./sudo.nix {};
+  sudo-rs = handleTest ./sudo.nix { sudo = pkgs.sudo-rs; };
```
In the test you would also need to do

````
{ pkgs, sudo ? pkgs.sudo, ... }: {
  ...
   nodes = {
     ...
       security.sudo.package = sudo;
   ...
   

```

nicoonoclaste committed:
nixos/sudo: Only wrap `sudoedit` when using Miller's sudo

mkg20001 committed:
nixos/test-driver: do not break if the command writes to stderr

Capturing `stderr` as part of the return `output` could break existing tests.

nicoonoclaste committed:
nixos/sudo: Generate `sudo-i` PAM config for interactive use of `sudo-rs`

nicoonoclaste committed:
nixos/sudo: Add myself as maintainer

mkg20001 commented on nixos/lib/test-driver/test_driver/machine.py:
```diff
@@ -582,9 +582,7 @@ def execute(
-        out_command = (
-            f"{timeout_str} bash -c {shlex.quote(command)} | (base64 -w 0; echo)\n"
-        )
+        out_command = f"{timeout_str} bash -c {shlex.quote(command)} 2>/dev/null | (base64 -w 0; echo)\n"
```
maybe we should add a flag for capturing stderr

mkg20001 reviewed (approved):
lgtm, I'd like another more sudo-related maintainer to look over aswell

delroth commented on nixos/tests/sudo.nix:
```diff
@@ -42,38 +46,70 @@ in
       };
 
     nodes.strict = { ... }: {
+      environment.systemPackages = [ pkgs.faketty ];
```
-1 from me on using faketty in nixos sudo tests. sudo has a bunch of complex logic regarding PTYs and I'm not confident this will be properly exercised anymore with faketty.

IMO if sudo-rs requires it, that's a bug in sudo-rs.

delroth commented on nixos/modules/security/sudo.nix:
```diff
@@ -258,12 +285,12 @@ in
-          # Make sure that the sudoers file is syntactically valid.
-          # (currently disabled - NIXOS-66)
-          "${pkgs.buildPackages.sudo}/sbin/visudo -f $src -c && cp $src $out";
+          "${cfg.package}/bin/visudo -f $src -c && cp $src $out";
```
Where did `buildPackages` go?

lopsided98 replied:
This breaks cross-compiling, `cfg.package` is for the host platform; this runs at build time which is why it used `buildPackages`

delroth reviewed (dismissed):
In general, I signed up to maintain sudo, not sudo-rs. I don't use sudo-rs, I'm not planning to, so I can't properly test future changes to unit tests and such.

Tying together the NixOS implementations and tests for sudo and sudo-rs means that maintainers for both need to care about both. That's currently not the case, at least from my side. To me, this is just extra complexity you're pushing onto me in the future.

If sudo-rs was properly compatible with what it claims to be replacing, maybe that would be a different story. As it is, it's not, and it shows in the size of the diff here.

delroth commented:
(Note: an acceptable resolution here would also be to drop me from the maintainers set for sudo, and add yourself instead.)

mkg20001 commented:
>As it is, it's not, and it shows in the size of the diff here.

To be fair, sudo-rs is fairly new and over time the diff will likely vanish completely

Nyabinary commented:
I'll be willing to help maintain sudo-rs as well, since I intend to use it.

Ma27 commented:
> Nobody is asking you to maintain sudo-rs?

I think the problem is that the sudo module is modified, so as a maintainer of sudo you'll have to always watch out for sudo-rs compatibility when touching that at least.

We had a similar issue with forgejo/gitea (which used to share a module) which eventually lead to incompatibilities with the former (and as a result it was decided to duplicate the module).

delroth commented:
Sorry, I have no energy to engage with this at the moment, especially after how dismissive I feel you've been of my prior comment on this PR. My concerns still stand.

Nyabinary commented:
Can we not dissolve into drama for this please lol

mkg20001 commented:
My two cents

Regarding faketty: sudo lecture was disabled due to the pty issue, which is now solved. it could possibly be enabled in another pr.

Regarding maintaince: the difference is meant to become less over time, except for explicitly missing features in sudo-rs, which aren't many. So we shouldn't split unless it gets considerably worse, as currently it's just SETENV being missing.

---


given how large this has become and the lack of other issues with this pr I'd go for a merge, we can do the rest after merge. Is this okay with everyone?

Ma27 commented:
> Can we not dissolve into drama for this please lol

This phrasing and then linking this into a Matrix channel is questionable at least.

> Considering that @delroth never committed to said module, that was presumably not the issue there.

I disagree with the metric: taking care of sudo means that you'd also need to touch the module _if_ necessary. Perhaps this wasn't the case in the time @delroth did that. However if that becomes necessary now, it's also necessary to check that sudo-rs is still fine.

I'm sure it wasn't meant that way by you, but especially the `Nobody is asking you to maintain sudo-rs?` seems pretty dismissive.

> The current situation is that there is no documented maintainer for the sudo module (as opposed to the drv you maintain)

Yeah, but the `meta.maintainers` is not really enforced on NixOS modules. RIght now maintaining a package implies in most of hte cases that said person also looks after the NixOS module.

I still think it'd be preferable to create a `security.sudo-rs` (or whatever namespace) which does all the `sudo-rs` specific stuff. As soon as it's an actual drop-in replacement, we can still merge the two together.

lopsided98 commented:
Mostly repeating my comment on the diff above to make it easier to find: https://github.com/NixOS/nixpkgs/pull/253876#discussion_r1332303128

This PR, specifically f0107b4, breaks cross-compiled NixOS because it no longer uses `buildPackages` and therefore tries to run host platform `visudo` as build time.

RaitoBezarius commented:
I am absolutely unhappy about the handling of this PR.

Changes to the test driver were done without waiting for the code owner (@tfc) on them.

This PR was insufficiently tested as cross-compilation was broken even though it was mentioned explicitly in https://github.com/NixOS/nixpkgs/pull/253876#discussion_r1327626076.

This PR was rushed and merged too fast given the open unresolved comments, I am inclined to revert to it, @nbraud.

Given the controversy around `sudo-rs` and its maintenance, I agree that `sudo-rs` should gain its own module in the time being as `sudo-rs` **is not** a 1:1 drop-in replacement of `sudo`, it contains edge cases and subtelties and I am not so sure you are promising here to deal with them moving forward.

Rather than dealing with more issues for the upcoming 23.11 release of NixOS, `sudo-rs` can only come via a new NixOS module or will have to wait after November, that's my personal take on the whole situation.

RaitoBezarius commented:
> I really do not believe this is the right place to have those conversations however (be it you or @Nyabinary) and I won't engage further on that topic.

Also, this is *the place* to have the discussion regarding `sudo(-rs)`, I am sorry, there is no other place as not everyone is expected to have a Matrix channel to discuss this and the discussion has to be public.

RaitoBezarius commented:
I opened the revert in https://github.com/NixOS/nixpkgs/pull/256472 as a draft, I am waiting for your inputs on how to sort out this situation.

mkg20001 commented:
I merged it as we can tweak things later, merging this doesn't break anything and it avoids having PRs die, as has happened often enough.

The specific test driver change (2>/dev/null) only changes stderr from being printed out via the console to not being printed out. It can be reverted individually, doesn't change test behavior and should not be used as an argument of "let's revert the whole thing". You can revert that one, I'm fine with it.

The changes in detail that I'm aware which are not just refactors of the module:
- One new pam service, behind an mkIf
- SETENV is not supported on sudo-rs so it's off when sudo-rs is detected

> it contains edge cases and subtelties and I am not so sure you are promising here to deal with them moving forward.

@nbraud added her to the maintainers for the module. That looks like she is.

This PR was open for two weeks. We explicitly solicited input on it. We got that. We did stuff, like later the tests got split, as requested. We insist on the module considering it really is a pain in the ass, considering the diff is already just two things and that's it.

The only open issue, at the time of merge, was the module. The tests got split. Feel free to bring up what wasn't resolved.

RaitoBezarius commented:
> I merged it as we can tweak things later, merging this doesn't break anything and it avoids having PRs die, as has happened often enough.

Just to be clear, it *broke* things and things which were predictably mentioned in the review process.

I understand the problem about having PRs to die but using merge privilege to force your way through is a risky strategy because then you have the community who needs to intervene.

> The specific test driver change (2>/dev/null) only changes stderr from being printed out via the console to not being printed out. It can be reverted individually, doesn't change test behavior and should not be used as an argument of "let's revert the whole thing". You can revert that one, I'm fine with it.

Yes, but there's an attitude problem here. We had two lines change in the test driver which caused pains because then someone has to triage what happened, and it does not help that CODEOWNERS are bypassed on those aspects because we lose the information.

> This PR was open for two weeks. We explicitly solicited input on it. We got that. We did stuff, like later the tests got split, as requested. We insist on the module considering it really is a pain in the ass, considering the diff is already just two things and that's it.

This may be a hot take and discouraging, but changes for something like sudo cannot take 2 weeks in my opinion, sudo is at the core of our systems and requires much more scrutiny and should have much more approvals. I know this is not easy, but this is the responsibility of the authors to drive the reviewing process and demand for inputs, there is no way around this otherwise we are just running into everyone does their thing and if they are not happy they just send it and whatever happens.

In 2 weeks, you got reactions, reviews and controversy, that's already a big reason for pausing and taking the time to garner more support for the changes.

Instead, it was sent.

The whole issue here is not really about "what are the issues in that PR?", this PR have been brought to me very early but because I have already too much to deal with, I decided to leave it to their authors and I would have been fine with anything as long as it respects a reasonable review process.

So please ensure that this (:= fast merges while discarding open comments and not taking the time to leave it open) never happens again. PRs dying is a problem, sure, but complicated core changes takes time, and we have to take the time, there is no need to go at speed of light for those changes in Nixpkgs. Remember that folks are *volunteers*.
